lib/finch/http1/pool.ex

```diff
@@ -38,6 +38,112 @@ defmodule Finch.HTTP1.Pool do
     )
   end
 
+  @impl Finch.Pool
+  def stream(pool, req, name, opts) do
+    pool_timeout = Keyword.get(opts, :pool_timeout, 5_000)
+    receive_timeout = Keyword.get(opts, :receive_timeout, 15_000)
+    request_timeout = Keyword.get(opts, :request_timeout, 30_000)
+
+    metadata = %{request: req, pool: pool, name: name}
+
+    start_time = Telemetry.start(:queue, metadata)
+    owner = self()
+    ref = make_ref()
+
+    holder =
+      spawn_link(fn ->
+        try do
+          NimblePool.checkout!(
```

Comment:
This either needs to happen when the stream starts or you need to check later on that the process that checked out is the one the one that is streaming. Otherwise someone will pass the stream around to another process and it won't behave as expected.

Reply:
> Otherwise someone will pass the stream around to another process and it won't behave as expected

I've tested it and it worked as expected, I'll push the test in a sec. Take a look at it please, perhaps I've misunderstood it.

Reply:
To be clear, it is probably best to do this lazily, if possible. As it is more flexible.

Reply:
I will be bouncing out soon, but I meant this:

    stream = Finch.actual_stream(fn -> ... end)
    spawn(fn -> Enum.to_list(stream) end)

Even if this works today, because the process doing the streaming is not the one linked to, you could run into situations where the connection is never checked backed in, such as this:

    stream = Finch.actual_stream(fn -> ... end)
    pid = spawn(fn -> Enum.each(stream, fn _ -> Process.sleep(:infinity) end)
    Process.sleep(1000)
    Process.exit(pid, :shutdown)

Reply:
> you could run into situations where the connection is never checked backed in, such as this:

Unfortunately, I don't see any way to track this without explicit links. This limitation can be reflected in the documentation. But it is generally true for any possible resourse-oriented stream, like `File.stream!/3`, so

1. No possible implicit solution
2. True for any resouse-stream

Therefore, I wouldn't take any action except documentation for this one

Reply:
That's not true. You can guarantee this won't leak resources by using the `after` callback of `Stream.resource` and linking. You just need to make sure that linking is done in the first argument of `Stream.resource`, so it is done against the process actually consuming the streaming.

Another issue with this implementation is that it checks out too early. Checking out should be done as soon as the stream begins. Doing that would solve both problems.

Reply:
> so it is done against the process actually consuming the streaming.

Yeah, right, I haven't thought about it

Reply:
Did it

Reply:
I have reviewed it and I like the new implementation a lot, it is much simpler. Great job! The only think remaining is making the checkout late and dealing with suspensions (which is used by `zip`). You have to be careful because a stream may not emit any item, be then suspended, and halted.

There are two ways you can do this: one is by moving the `after` callback, aka `send(holder, {ref, :stop, conn})`, to HTTP.conn. The flow of the code would be something like:


```
def stream(...) do
  # ...
  fn tagged_acc, fun ->
    conn = NimblePool.checkout!(...)

    HTTP1.request(tagged, fun, fn -> send(holder, {ref, :stop, conn}) end)
  end
end
```

Moving all error flow to HTTP.conn is simpler, because suspending would then just be something like this:

```
def request(..., {:suspend, acc}, function, after_fun) do
  {:suspended, acc, &request(..., &1, function, after_fun)}
end
```

In other words, streams are easier to implement if they are fully tail recursive, and not relying on try/catch and similar. Instead wrap each invocation of `fun` in try/catch.

Reply:
I've thought about it, and now we're back to the beginning where I've implemented it with `Stream.resourse`, but now I have to implement all of this suspend/halt logic plus change the request implementation

Reply:
Not really. The work is almost all done. Making it tail recursive is a matter of passing an after block to the existing code and adding one single clause to handle suspend. I still think it will be less code than the original PR. :)